### DISCUSSION TITLE
Dark mode: system vs web app settings

### DISCUSSION THREAD

**Discussion by husseinalhammad** [email-markup-consortium/email-markup-consortium#99] [General]
While there is a way in CSS to detect the user's colour scheme preference as set in the Operating System (OS) or the user agent (browser), there is no native CSS way to check the user's preference on a web app which may not match what they have set as their preference in the OS or user agent.

This means using the CSS `prefers-color-scheme` media feature isn't enough to cover all cases.

Email developers would love to be able to author CSS for dark mode by simply relying on the CSS `prefers-color-scheme` media feature:

```css
@media (prefers-color-scheme: dark) {
  // dark mode styles
}
```

But the reality is if the user has the option to set their preferred colour scheme through the web app, then the CSS `prefers-color-scheme` media feature is simply not enough as it won't be able to query the user's preference on the web app level.

Websites that give the user the option to choose a colour scheme often give 3 options:
- Light
- Dark
- System (whatever the OS or user agent is set to)

The above applies to some webmail clients like Outlook web:
![Screenshot of Outlook web settings showing theme options: light, dark, system](https://github.com/user-attachments/assets/eb0ca7f7-f4b1-4f84-8a6b-27e9361c1996)

The CSS `prefers-color-scheme` media feature really only gives us the System preference. How should email clients make it possible for email developers to target different colour scheme preferences with only CSS?

**Comment by M-J-Robbins** [comment]
One solution is using a combination of [`color-scheme`](https://developer.mozilla.org/en-US/docs/Web/CSS/color-scheme) and the [`light-dark()`](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/light-dark) colour function.

In the below example, I'm using `light-dark` to set colours for both light and dark in a single property. 
When I change the `color-scheme` on the container it will switch between the colours. 

The first one is set to `dark` so will always show the dark colours.
The second one is set to `light` so will always show the light colours.
The third one is set to `light dark` so it will show either depending on the OS or user agent preference.

```html
<div class="test1">
  <div class="email">
    <h1>test</h1>
  </div>
</div>
<div class="test2">
  <div class="email">
    <h1>test</h1>
  </div>
</div>
<div class="test3">
  <div class="email">
    <h1>test</h1>
  </div>
</div>
<style>
  .test1{
    color-scheme:dark;
  }
  .test2{
    color-scheme:light;
  }
  .test3{
    color-scheme:light dark;
  }
  .email{
    background:light-dark(#fed, #654);
    color:light-dark(#123, #def);
  }
</style>
```

**Reply by husseinalhammad** [reply]
This is a really nice approach if we're only concerned with color values, but this doesn't allow a developer to make use of the CSS `prefers-color-scheme` media feature for other things like:
1. making other styling changes (e.g. darken an image with a CSS filter)
2. use it for art direction in `<picture>` element

**Reply by o-t-w** [reply]
IMO it would be nicer for email clients to adopt an approach that utilises standard web browser features. Compared to how difficult it is to style dark mode in email at the moment, even a limited approach would be a huge improvement.

**Reply by o-t-w** [reply]
This sounds like it would be the ideal solution: https://drafts.csswg.org/mediaqueries-5/#script-control-user-prefs

**Comment by M-J-Robbins** [comment]
Another option is what Mail.ru does. 
It will replace `@media (prefers-color-scheme: light)` or `@media (prefers-color-scheme: dark)` with `@media (min-height: 0px)` depending on the app color scheme so that the styes are applied in the message.

**Reply by husseinalhammad** [reply]
This sounds like a reasonable approach. 

Are there any potential side effects? I can see a developer using `@media (min-height: 0px)` to exclude Outlook on Windows (classic version):

```css
@media (min-height: 0px) {
  // styles
}
```

Is there a chance the email client would mistakenly remove or rewrite this? Is there a more foolproof way with a smaller risk of something like this happening.

**Reply by M-J-Robbins** [reply]
I think i read somewhere recently that there is a proposal for allowing CSS custom properties inside media queries (can't find the link currently). I'm wondering if that might help a little here. 

So the email clients would convert the prefers-color-scheme queries to`@media (prefers-color-scheme: var(--light))` and `@media (prefers-color-scheme: var(--dark))`. Makes it easier to read in dev tools too.

**Reply by M-J-Robbins** [reply]
Or possibly better still would be to use `@scope`. The email client could add a class wrapping the content of `light` or `dark`. Then replace `@media (prefers-color-scheme: dark){}` with `@scope(.dark){}`.  Currently not supported in Firefox but works elsewhere.
